Задача решена

z0rrus opened the pull request.

z0rrus committed:
Задача решена

ac1d0globlya commented on src/main/java/ru/sbt/jschool/session2/OutputFormatter.java:
```diff
@@ -1,34 +1,157 @@
+        }
+    }
+
+    private Class<?> inferColumnType(Object[][] data, int colIndex) {
```
Давай сделаем 

1. интерфейс вида

```
interface Printer {
    List<Class<?>> supported();
    int length(Object obj);
    String print(Object obj);
}
```

2. для каждого типа сделаем реализацию `StringPrinter`, `DatePrinter`, `DoublePrinter`, `NumberPrinter`.

3. Сделаем `Map<Class<?>, Printabler> knownPrinters` в которую положим реализации для каждого класса который поддерживается (возвращается методов supported).

Тогда код печати упростится до

```java
private final Printer anyPrinter = new Printer() {
    public void List supported() {
         return Collections.emptyList();
    }
    public String print(Object obj) {
         return Objects.toString(obj);
    }
    public int length(Object obj) {
         return obj == null ? 4 : print(obj).size();
    }
}

Printer print(Object obj) {
    if (obj == null) {
         return anyPrinter;
    }

    return knownPrinters.getOrDefault(obj.getClass(), anyPrinter);
}

```

z0rrus committed:
Исправленный вариант

Учтены исправления в коде, добавлен наглядный вывод в консоль для OutputFormaterTest

nizhikov commented on src/main/java/ru/sbt/jschool/session2/OutputFormatter.java:
```diff
@@ -1,34 +1,226 @@
+        @Override
+        public List<Class<?>> supported() { return Collections.singletonList(Date.class); }
+        @Override
+        public int length(Object obj) { return obj == null ? 1 : print(obj).length(); }
```
Здесь мы точно знаем длину записи. Она постоянная. Нет смысла каждый раз print вызывать.

nizhikov commented on src/main/java/ru/sbt/jschool/session2/OutputFormatter.java:
```diff
@@ -1,34 +1,226 @@
+            colWidths[c] = names[c].length();
+        }
+
+        String[][] formattedData = new String[rowCount][colCount];
```
Зачем хранить данные?
Если строк будет 1_000_000 то будем все хранить перед печатью?

nizhikov commented on src/main/java/ru/sbt/jschool/session2/OutputFormatter.java:
```diff
@@ -1,34 +1,226 @@
+                Printer printer = colPrinters[c];
+                String cell = formattedData[r][c];
+                // Left-align text, right-align numbers/dates
+                if (printer instanceof StringPrinter) {
```
Знания о том куда "прижимать" вывод нужно хранить в Printer'е.
Т.е в него добавить метод
```
default boolean leftAlign() { return false; }
```

nizhikov commented on src/main/java/ru/sbt/jschool/session2/OutputFormatter.java:
```diff
@@ -1,34 +1,226 @@
+                String formatted = colPrinters[c].print(data[r][c]);
+                formattedData[r][c] = formatted;
+                if (formatted.length() > colWidths[c]) {
+                    colWidths[c] = formatted.length();
```
Здесь надо испольозовать `printer.length()`

z0rrus committed:
16.07
